assume.go

```diff
@@ -0,0 +1,109 @@
+// Copyright William Schwartz 2014. See the LICENSE file for more information.
+
+package pigosat
+
+// #include "picosat.h"
+import "C"
+import "unsafe"
+
+// For testing litArrayToSlice in TestlitArrayToSlice since you can't import CGo
+// into test files.
+var cArray123 = [3]C.int{1, 2, 3}
+var cArray1230 = [4]C.int{1, 2, 3, 0}
+var cZero C.int = 0
+
+// Assume adds a temporary unit clause, i.e., a clause containing the one
+// literal you pass as an argument. An assumption remains valid after the next
+// call to Solve returns until a call to AddClauses, Assume, or a second Solve.
+// You can add arbitrary many assumptions before the next call to Solve.
+// Methods FailedAssumptions, FailedAssumptions, MaxSatisfiableAssumptions, and
+// NextMaxSatisfiableAssumptions operate on the current, valid assumptions.
+func (p *Pigosat) Assume(lit Literal) {
+	defer p.ready(false)()
+	p.couldHaveFailedAssumptions = false
+	// void picosat_assume (PicoSAT *, int lit);
+	C.picosat_assume(p.p, C.int(lit))
+}
+
+// FailedAssumption returns whether a given literal is a valid, "failed"
+// assumption (see Assume), meaning the last call to Solve used the assumption
+// to derive unsatisfiability. FailedAssumption always returns false when the
+// last call to Solve had status other than Unsatisfiable. This may be an
+// over-approximation of the set of assumptions necessary to derive
+// unsatisfiability.
+func (p *Pigosat) FailedAssumption(lit Literal) bool {
+	defer p.ready(true)()
+	// picoast_failed_assumption SIGABRTs if the following conditional is true
+	if p.res() != Unsatisfiable || lit == 0 || !p.couldHaveFailedAssumptions {
+		return false
+	}
+	// int picosat_failed_assumption (PicoSAT *, int lit);
+	return C.picosat_failed_assumption(p.p, C.int(lit)) != 0
+}
+
+// litArrayToSlice converts a 0-terminated C array of ints (of length at most
+// maxLen) to a Go slice of Literals.
+func litArrayToSlice(litPtr *C.int, maxLen int) []Literal {
+	if litPtr == nil {
+		panic("NULL pointer")
+	}
+	lits := []Literal{}
+	for *litPtr != 0 {
+		if len(lits) >= maxLen {
+			panic("Array not zero-terminated")
```

Comment:
`"Array not zero-terminated"` isn't the best error message there. My idea is this, keeping in mind that `litArrayToSlice` is an internal utility function. We only use `litArrayToSlice` on zero-terminated arrays coming from PicoSAT. In general, these arrays are zero-terminated. In case for whatever reason they are not, we can pass in `maxLen`, which we known in general is equal to `p.Variables()`, so that we don't loop forever here. In other words, `maxLen` is purely a safety control and we don't know how long the array actually is until we hit the zero terminator.

Thoughts on a better error message?